src/xtgeo/io/_file_wrapper.py

```diff
@@ -0,0 +1,618 @@
+"""A FileWrapper class to wrap around files and streams representing files."""
+from __future__ import annotations
+
+import io
+import os
+import pathlib
+import platform
+import re
+import struct
+import uuid
+from os.path import join
+from tempfile import mkstemp
+from typing import TYPE_CHECKING, Literal, Union
+
+import h5py
+
+import xtgeo._cxtgeo
+from xtgeo.common import null_logger
+
+logger = null_logger(__name__)
+
+if TYPE_CHECKING:
+    import pathlib.Path
+
+    from xtgeo.cube import Cube
+    from xtgeo.grid3d import Grid, GridProperties, GridProperty
+    from xtgeo.surface import RegularSurface, Surfaces
+    from xtgeo.wells import BlockedWell, BlockedWells, Well, Wells
+    from xtgeo.xyz import Points, Polygons
+
+    XTGeoObject = Union[
+        BlockedWell,
+        BlockedWells,
+        Cube,
+        Grid,
+        GridProperty,
+        GridProperties,
+        Points,
+        Polygons,
+        RegularSurface,
+        Surfaces,
+        Well,
+        Wells,
+    ]
+
+SUPPORTED_FORMATS = {
```

Comment:
I think a dataclass would be nice here.

Written by GPT:

```python
from dataclasses import dataclass, field
from typing import Literal


@dataclass
class SupportedFormats:
    rmswell: list[
        Literal[
            "rmswell",
            "rmsw",
            "w",
            "bw",
        ]
    ] = field(
        default_factory=lambda: [
            "rmswell",
            "rmsw",
            "w",
            "bw",
        ]
    )
    roff_binary: list[
        Literal[
            "roff_binary",
            "roff",
            "roff_bin",
            "roff-bin",
            "roffbin",
            "roff.*",
        ]
    ] = field(
        default_factory=lambda: [
            "roff_binary",
            "roff",
            "roff_bin",
            "roff-bin",
            "roffbin",
            "roff.*",
        ]
    )
    roff_ascii: list[
        Literal[
            "roff_ascii",
            "roff_asc",
            "roff-asc",
            "roffasc",
            "asc",
        ]
    ] = field(
        default_factory=lambda: [
            "roff_ascii",
            "roff_asc",
            "roff-asc",
            "roffasc",
            "asc",
        ]
    )
    egrid: list[Literal["egrid",]] = field(
        default_factory=lambda: [
            "egrid",
        ]
    )
    fegrid: list[Literal["fegrid",]] = field(
        default_factory=lambda: [
            "fegrid",
        ]
    )
    init: list[Literal["init",]] = field(
        default_factory=lambda: [
            "init",
        ]
    )
    finit: list[Literal["finit",]] = field(
        default_factory=lambda: [
            "finit",
        ]
    )
    unrst: list[Literal["unrst",]] = field(
        default_factory=lambda: [
            "unrst",
        ],
    )
    funrst: list[Literal["funrst"]] = field(
        default_factory=lambda: [
            "funrst",
        ]
    )
    grdecl: list[Literal["grdecl"]] = field(
        default_factory=lambda: [
            "grdecl",
        ],
    )
    bgrdecl: list[Literal["bgrdecl"]] = field(
        default_factory=lambda: [
            "bgrdecl",
        ]
    )
    irap_binary: list[
        Literal[
            "irap_binary",
            "irap_bin",
            "rms_binary",
            "irapbin",
            "gri",
        ],
    ] = field(
        default_factory=lambda: [
            "irap_binary",
            "irap_bin",
            "rms_binary",
            "irapbin",
            "gri",
        ]
    )
    irap_ascii: list[
        Literal[
            "irap_ascii",
            "irap_asc",
            "rms_ascii",
            "irapasc",
            "fgr",
        ],
    ] = field(
        default_factory=lambda: [
            "irap_ascii",
            "irap_asc",
            "rms_ascii",
            "irapasc",
            "fgr",
        ]
    )
    hdf: list[
        Literal[
            "hdf",
            "hdf5",
            "h5",
        ]
    ] = field(
        default_factory=lambda: [
            "hdf",
            "hdf5",
            "h5",
        ]
    )
    segy: list[
        Literal[
            "segy",
            "sgy",
            "segy.*",
        ]
    ] = field(
        default_factory=lambda: [
            "segy",
            "sgy",
            "segy.*",
        ]
    )
    storm: list[Literal["storm"]] = field(
        default_factory=lambda: [
            "storm",
        ]
    )
    zmap_ascii: list[
        Literal[
            "zmap",
            "zmap+",
            "zmap_ascii",
            "zmap-ascii",
            "zmap-asc",
            "zmap.*",
        ]
    ] = field(
        default_factory=lambda: [
            "zmap",
            "zmap+",
            "zmap_ascii",
            "zmap-ascii",
            "zmap-asc",
            "zmap.*",
        ]
    )
    ijxyz: list[Literal["ijxyz"]] = field(
        default_factory=lambda: [
            "ijxyz",
        ]
    )
    petromod: list[
        Literal[
            "pmd",
            "petromod",
        ]
    ] = field(
        default_factory=lambda: [
            "pmd",
            "petromod",
        ]
    )
    xtg: list[
        Literal[
            "xtg",
            "xtgeo",
            "xtgf",
            "xtg.*",
        ]
    ] = field(
        default_factory=lambda: [
            "xtg",
            "xtgeo",
            "xtgf",
            "xtg.*",
        ]
    )
    xyz: list[
        Literal[
            "xyz",
            "poi",
            "pol",
        ]
    ] = field(
        default_factory=lambda: [
            "xyz",
            "poi",
            "pol",
        ]
    )
    rms_attr: list[
        Literal[
            "rms_attr",
            "rms_attrs",
            "rmsattr.*",
        ]
    ] = field(
        default_factory=lambda: [
            "rms_attr",
            "rms_attrs",
            "rmsattr.*",
        ]
    )
```

Reply:
I played around with how this would work and I'm not sure a dataclass is the way to go since these values are static. Would have to go from

```python
for fmt, variants in SUPPORTED_FORMATS.items():
    if suffix in variants:
        logger.debug("Extension hints: %s", fmt)
        return fmt
```

to

```python
supported_formats = SupportedFormats()
for fmt in fields(supported_formats):
    if suffix in getattr(supported_formats, fmt.name):
        logger.debug("Extension hints: %s", fmt)
        return fmt.name
```

and making new instances of the dataclass wherever these values are used. Not a terrible exchange, but with default factory fields in tandem it feels rather like working around dataclasses than with them.

I made #1129 which I think conceptually captures what you're suggesting